torch/backends/cudnn/__init__.py

```diff
@@ -1,61 +1,27 @@
-import os
-import ctypes
 import sys
 import torch
 import warnings
-from torch.version import cuda
 from contextlib import contextmanager
-from subprocess import Popen, PIPE
 from torch.backends import ContextProp, PropModule, __allow_nonbracketed_mutation
 
+try:
+    from torch._C import _cudnn
+except ImportError:
+    _cudnn = None
+
 # Write:
 #
 #   torch.backends.cudnn.enabled = False
 #
 # to globally disable CuDNN
 
-lib = None
 __cudnn_version = None
-# TODO: dynamic version checks via cudnnGetVersion
-
-def find_cudnn_windows_lib():
-    # Override the default search process
-    # Fixes https://github.com/pytorch/pytorch/issues/20202
-    # The library selection will be done in these directories one by one
-    # 1. [Package Root]\Lib 
-    #    That's where our libraries are in, which should be loaded first.
-    # 2. Default directories
-    #    That is stored in the environment variable `PATH`.
-    test_env = os.environ.copy()
-    old_path = test_env['PATH']
-    th_dll_path = os.path.join(os.path.dirname(
-        os.path.dirname(os.path.dirname(__file__))), 'lib')
-    test_env['PATH'] = ';'.join([th_dll_path, old_path])
-    proc = Popen(['where', 'cudnn64*.dll'], stdout=PIPE,
```

Comment:
There's no RPATH for Windows, so we will have to depend on the user's environment. If a user compiled PyTorch by himself and the CUDA libraries are not in `PATH`, then the user will encounter a DLL load failure.

Reply:
@peterjc123 I suppose what I don't understand, is prior to hitting the code here, we will have loaded `_C.dll` into the process. This indirectly depends on `cudnn64.dll`. So how come this code works today? (Even if later, when we load the libraries with ctypes, we have to do all this faffing about, it doesn't seem like it would apply here.)

Reply:
Yes, I'm wrong here. Currently, it is using `PATH` and we didn't add any guess logic before loading `_C.dll`. But in Python 3.8, they changed this behaviour. Please see https://github.com/pytorch/pytorch/pull/28536/files#r357489428 for more details.

Reply:
Taking a look at the binary packaging for 1.3.1, it looks like `cudnn64.dll` is distributed with PyTorch, and I guess (????) that you're assumed to have put the CUDA runtime into your PATH (certainly, conda activate + cudatoolkit will get it in your PATH)

Reply:
Yes, that's true. For binary builds, it is getting into `PATH` through
1. Conda package: Relies on `cudatoolkit`, which it will be added into PATH by conda itself.
2. Pip package: The CUDA runtime DLLs are copied into [PY_LIB_DIR]/torch/lib and then we add that dir into PATH.
But for developers that build PyTorch on their own, it will directly depend on their setting of `PATH`. But generally speaking, it should be added during the installation of CUDA.